### DISCUSSION TITLE
Traits for layouts

### DISCUSSION THREAD

**Discussion by ShaikhDanish04** [livewire/livewire#3390] [1. Help]
Here we are trying to create a boilerplate with some constraints to easily create layouts for new people who join our community with laravel8-livewire2-bootstrap5 but we are having an issue in creating traits for the layout, here the components **render** function is invoked before the **rendered hook** of trait which is not allowing the trait to append custom layouts or maybe some else is happening.

Can we get some help to find the solution on it or else the only way to use the layouts on full-page components is via naming layouts or extends in each file:

```
  function render()
    {
        return view('livewire.pages.dashboard')->layout('layouts.panel');
    }
```

### Dashboard

```
namespace App\Http\Livewire\Pages;

use App\Http\Livewire\Traits\WithPanelLayout;
use Livewire\Component;

class Dashboard extends Component
{
    use WithPanelLayout;

    function render()
    {
        return view('livewire.pages.dashboard');
    }
}
```
### WithPanelLayout
```
namespace App\Http\Livewire\Traits;

trait WithPanelLayout
{
    public function RenderedWithPanelLayout($view)
    {
        $view->layout('layout.panel');
    }
}
```

**Comment by Dimonka2** [comment]
You cannot usually have several layouts in a component, thus another option could be to use layout inheritance:

```php
class Dashboard extends ComponentWithPanelLayout
{
protected function getView(){...}
```

**Reply by ShaikhDanish04** [reply]
Can't we have a solution in that way because we have to make it easy?

**Reply by Dimonka2** [reply]
Of cause you can make your own base component that supports this type of traits. 
And if you need it in several projects, you can pack it in your own package.

**Reply by ShaikhDanish04** [reply]
Can you share some guidance on how to implement it?

**Reply by Dimonka2** [reply]
Easiest way is to make a component that has a render function which expects to have a view and a layout.
For example if you put the rendering logic in your components into `getView()` function it will look like this:
```php
abstract class ComponentWithLayout
{
     protected $layout;
     
     public function render()
     {
         $view = $this->getView();
         if($this->layout) $view->layout($this->layout);
         return $view;
     }
}
```
Your trait would be like this
```php
namespace App\Http\Livewire\Traits;

trait WithPanelLayout
{
    public function initializeWithPanelLayout()
    {
        $this->layout = 'layout.panel';
    }
}
```
Of cause if you want to pass variables to layout it needs to use a reflection and loop through traits, check if there is a function `Rendered%TraitName%` and call it with a view as parameter. But anyway it has nothing to do with Livewire itself.
